cmd/infra/aws/create.go

```diff
@@ -1,225 +1,394 @@
+		OIDCIngressRoleArn:       getStackOutput(stack, "OIDCIngressRoleArn"),
+		OIDCImageRegistryRoleArn: getStackOutput(stack, "OIDCImageRegistryRoleArn"),
+		OIDCCSIDriverRoleArn:     getStackOutput(stack, "OIDCCSIDriverRoleArn"),
+		OIDCIssuerURL:            getStackOutput(stack, "OIDCIssuerURL"),
+		OIDCBucketName:           getStackOutput(stack, "OIDCBucketName"),
 	}
-	igwID, err := o.CreateInternetGateway(client, result.VPCID)
-	if err != nil {
-		return nil, err
+
+	return output, nil
+}
+
+func (o *CreateInfraOptions) createStack(ctx context.Context, cf *cloudformation.CloudFormation, baseDomainZoneID string) (*cloudformation.Stack, error) {
+	createStackInput := &cloudformation.CreateStackInput{
+		Capabilities: []*string{aws.String(cloudformation.CapabilityCapabilityNamedIam)},
+		TemplateBody: &hypercf.ClusterTemplate,
+		StackName:    aws.String(o.InfraID),
+		Tags: []*cloudformation.Tag{
+			{
+				Key:   aws.String("hypershift.openshift.io/infra"),
+				Value: aws.String("owned"),
+			},
+		},
+		Parameters: []*cloudformation.Parameter{
+			{
+				ParameterKey:   aws.String("InfrastructureName"),
+				ParameterValue: aws.String(o.InfraID),
+			},
+			{
+				ParameterKey:   aws.String("BaseDomainHostedZoneId"),
+				ParameterValue: aws.String(baseDomainZoneID),
+			},
+			{
+				ParameterKey:   aws.String("Subdomain"),
+				ParameterValue: aws.String(o.Subdomain),
+			},
+		},
 	}
-	natGatewayID, err := o.CreateNATGateway(client, result.PublicSubnetID, result.Zone)
-	if err != nil {
-		return nil, err
+
+	if !o.PreserveOnFailure {
+		createStackInput.OnFailure = aws.String(cloudformation.OnFailureDelete)
 	}
-	result.SecurityGroupID, err = o.CreateWorkerSecurityGroup(client, result.VPCID)
+
+	createStackOutput, err := cf.CreateStack(createStackInput)
 	if err != nil {
 		return nil, err
 	}
-	privateRouteTable, err := o.CreatePrivateRouteTable(client, result.VPCID, natGatewayID, result.PrivateSubnetID, result.Zone)
+
+	log.Info("Waiting for infrastructure to be created", "id", o.InfraID, "stackID", *createStackOutput.StackId)
+	var stack *cloudformation.Stack
+	err = wait.PollUntil(5*time.Second, func() (bool, error) {
+		latest, err := getStack(cf, *createStackOutput.StackId)
+		if err != nil {
+			log.Error(err, "failed to get stack", "id", *createStackOutput.StackId)
+			return false, nil
+		}
+		stack = latest
+		switch *stack.StackStatus {
+		case cloudformation.StackStatusCreateComplete:
+			return true, nil
+		case cloudformation.StackStatusCreateInProgress:
+			return false, nil
+		case cloudformation.StackStatusCreateFailed:
+			return false, fmt.Errorf("stack creation failed")
+		case cloudformation.StackStatusRollbackInProgress,
+			cloudformation.StackStatusRollbackComplete,
+			cloudformation.StackStatusRollbackFailed:
+			return false, fmt.Errorf("stack creation failed and was rolled back")
+		default:
+			return false, fmt.Errorf("unexpected stack creation status: %s", *stack.StackStatus)
+		}
+	}, ctx.Done())
```

Comment:
this is great now but still the error returned here will be a meaningless timeout message returned by `PollUntil`, it'd be useful to capture the one returned by your inlined function and pass it though here, something like:
```
var actualErr
err = wait.PollUntil(5*time.Second, func() (bool, error) {
  actualErr = 
} ctx.Done())
	if err != nil {
		return nil, fmt.Errorf("error getting stack: %v, %w", err, actualErr)
	}
```

wdyt?

Reply:
If `PollUtil` returns because of an error I returned inline (not a timeout), the useful error is propagated out already. If `PollUntil` returns because of the context closing (eg timeout), the timeout error generated by `PollUntil` is accurate, because the function didn't return due to the inline code returning any error — there is no actual error to wrap. Make sense? Maybe I'm misunderstanding.

Reply:
right, sorry I was forgetting and returning error finish the PollUntil right away.